# Let's take it up to Eleven

Now that we're familiar with the basics of data cleaning and analysis in pandas, we're going to take it up a notch. The U.S. Census fortunately makes their data available in a clean and straightforward format. However, in many cases, the data we receive can be extremely messy and difficult to manage.

That's why we're going to practice with a more unweildy. You can download the data [here](https://s3.amazonaws.com/nicar15/njaccidents.csv). You'll notice that it's quite a big file– about 1.7 million rows! These are reports from accidents in New Jersey between 2008 and 2013 from the New Jersey Department of Transportation. The data was originally scraped from [PDFs of crash reports](http://www.state.nj.us/transportation/refdata/accident/) filled out by clerk.

Import pandas and let's load in our new and very messy data

In [20]:
import pandas as pd
njaccidents = pd.read_csv('njaccidents.csv')

You may notice that you get this warning.

"DtypeWarning: Columns (6,17) have mixed types. Specify dtype option on import or set low_memory=False. interactivity=interactivity, compiler=compiler, result=result)"

This dtype error happens when when a column has both strings and integer values. You can ignore this for now because we'll fix it soon. Open up the first few rows of our dataframe.

In [21]:
njaccidents.head()

case code   County Name         Municipality Name  \
0  2008010108-026816                ATLANTIC      ABSECON CITY               
1  2008010108-163190                ATLANTIC      ABSECON CITY               
2  2008010108-24779                 ATLANTIC      ABSECON CITY               
3  2008010108-3901                  ATLANTIC      ABSECON CITY               
4  2008010108-5016                  ATLANTIC      ABSECON CITY               

   Crash Date  Crash Day Of Week  Crash Time  Police Dept Code  \
0  03/04/2008                 TU        1539                01   
1  12/19/2008                 F         1114                01   
2  11/25/2008                 TU        0345                99   
3  03/31/2008                 M         0105                01   
4  01/25/2008                 F         0942                01   

           Police Department   Police Station   Total Killed  \
0  ATLANTIC CITY              AIU                          0   
1  ATLANTIC CITY              TRAFFIC                      0   
2  NJ TRANSIT P.D.            ATLANTIC CITY                0   
3  EAST WINDSOR               TRAFFIC UNIT                 0   
4  EGG HARBOR TWP             HQ                           0   

          ...            Is Ramp    Ramp To/From Route Name  \
0         ...                                                 
1         ...                                                 
2         ...                                                 
3         ...                                                 
4         ...                                                 

    Ramp To/From Route Direction  Posted Speed  Posted Speed Cross Street  \
0                                           50                              
1                                           50                              
2                                           10                         25   
3                                            0                              
4                                           50                         40   

   Latitude  Longitude  Cell Phone In Use Flag  \
0  39.41158   74.49162                       N   
1  39.39231   74.48952                       N   
2                                            N   
3                                            N   
4  39.43036   74.52469                       N   

                               Other Property Damage  Reporting Badge No.  
0  NONE                                          ...                384    
1  NONE                                          ...                739    
2  ?                                             ...                0053   
3  NONE                                          ...                551    
4  NONE                                          ...                1571   

[5 rows x 47 columns]

Let's found out what we're working with, and get the column headers for all of the columns.

In [129]:
njaccidents.columns

Index([u'case code', u'County Name', u'Municipality Name', u'Crash Date',
       u'Crash Day Of Week', u'Crash Time', u'Police Dept Code',
       u'Police Department', u'Police Station', u'Total Killed',
       u'Total Injured', u'Pedestrians Killed', u'Pedestrians Injured',
       u'Severity', u'Intersection', u'Alcohol Involved', u'HazMat Involved',
       u'Crash Type Code', u'Total Vehicles Involved', u'Crash Location',
       u'Location Direction', u'Route', u'Route Suffix',
       u'SRI (Std Rte Identifier)', u'MilePost', u'Road System',
       u'Road Character', u'Road Surface Type', u'Surface Condition',
       u'Light Condition', u'Environmental Condition', u'Road Divided By',
       u'Temporary Traffic Control Zone', u'Distance To Cross Street',
       u'Unit Of Measurement', u'Directn From Cross Street',
       u'Cross Street Name', u'Is Ramp', u'Ramp To/From Route Name',
       u'Ramp To/From Route Direction', u'Posted Speed',
       u'Posted Speed Cross Street', u'Latitude

Oh boy. There's our first problem. Notice that there's a leading space in every column header. We should take it out.

In [175]:
njaccidents.rename(columns=lambda x: x.strip(), inplace=True)

Remember earlier where we renamed the columns in our dataframe? This time, we're using the same rename function to do take out all of the leading spaces using strip(). Pythonistas will notice that we're using the lambda python to apply strip() to every single column header. 

In [176]:
njaccidents.columns

Index([u'case code', u'County Name', u'Municipality Name', u'Crash Date',
       u'Crash Day Of Week', u'Crash Time', u'Police Dept Code',
       u'Police Department', u'Police Station', u'Total Killed',
       u'Total Injured', u'Pedestrians Killed', u'Pedestrians Injured',
       u'Severity', u'Intersection', u'Alcohol Involved', u'HazMat Involved',
       u'Crash Type Code', u'Total Vehicles Involved', u'Crash Location',
       u'Location Direction', u'Route', u'Route Suffix',
       u'SRI (Std Rte Identifier)', u'MilePost', u'Road System',
       u'Road Character', u'Road Surface Type', u'Surface Condition',
       u'Light Condition', u'Environmental Condition', u'Road Divided By',
       u'Temporary Traffic Control Zone', u'Distance To Cross Street',
       u'Unit Of Measurement', u'Directn From Cross Street',
       u'Cross Street Name', u'Is Ramp', u'Ramp To/From Route Name',
       u'Ramp To/From Route Direction', u'Posted Speed',
       u'Posted Speed Cross Street', u'Latitude

Great. Let's describe() the dataframe.

In [177]:
njaccidents.describe()

Total Killed   Total Injured  Pedestrians Killed  \
count  1707841.000000  1707841.000000      1707841.000000   
mean         0.001968        0.308403            0.000489   
std          0.047111        0.698354            0.022238   
min          0.000000        0.000000            0.000000   
25%          0.000000        0.000000            0.000000   
50%          0.000000        0.000000            0.000000   
75%          0.000000        0.000000            0.000000   
max          5.000000       42.000000            2.000000   

       Pedestrians Injured  Total Vehicles Involved     Road System  \
count       1707841.000000           1707841.000000  1707841.000000   
mean              0.016251                 1.878721        5.191971   
std               0.130033                 0.538178        2.499018   
min               0.000000                 0.000000        0.000000   
25%               0.000000                 2.000000        2.000000   
50%               0.000000                 2.000000        5.000000   
75%               0.000000                 2.000000        7.000000   
max              10.000000                20.000000       10.000000   

         Posted Speed  
count  1707840.000000  
mean        31.205212  
std         17.872276  
min          0.000000  
25%         25.000000  
50%         30.000000  
75%         45.000000  
max         99.000000

But let's see if we could describe() a column. Let's use the describe() function for the 'County Name' column header.

In [178]:
njaccidents['County Name'].describe()

count       1707841
unique           21
top       MIDDLESEX
freq         176402
Name: County Name, dtype: object

Interesting. So there are 21 unique values in the 'County Name' column (for the 21 counties in New Jersey). We can see that the top county with the most rows is Middlesex County with 176,402 crashes. What are the names of the counties in New Jersey? Let's find out by using the unique() function on our 'County Name' column.

In [179]:
njaccidents['County Name'].unique()

array(['ATLANTIC', 'BERGEN', 'BURLINGTON', 'CAMDEN', 'CAPE MAY',
       'CUMBERLAND', 'ESSEX', 'GLOUCESTER', 'HUDSON', 'HUNTERDON',
       'MERCER', 'MIDDLESEX', 'MONMOUTH', 'MORRIS', 'OCEAN', 'PASSAIC',
       'SALEM', 'SOMERSET', 'SUSSEX', 'UNION', 'WARREN'], dtype=object)

Well, that's frustrating. Looks like we're going to need to strip out the spaces out of the these county values. This time we'll use the map() function which will strip the white space out of every string found in the column.

In [180]:
njaccidents['County Name']=njaccidents['County Name'].map(str.strip)

Let's test.

In [181]:
njaccidents['County Name'].unique()

array(['ATLANTIC', 'BERGEN', 'BURLINGTON', 'CAMDEN', 'CAPE MAY',
       'CUMBERLAND', 'ESSEX', 'GLOUCESTER', 'HUDSON', 'HUNTERDON',
       'MERCER', 'MIDDLESEX', 'MONMOUTH', 'MORRIS', 'OCEAN', 'PASSAIC',
       'SALEM', 'SOMERSET', 'SUSSEX', 'UNION', 'WARREN'], dtype=object)

Perfect. Speaking of strings, let's fix that dtype error we got at the beginning of the exercise. Type in dtypes at the end of our dataframe.

In [182]:
njaccidents.dtypes

case code                          object
County Name                        object
Municipality Name                  object
Crash Date                         object
Crash Day Of Week                  object
Crash Time                         object
Police Dept Code                   object
Police Department                  object
Police Station                     object
Total Killed                        int64
Total Injured                       int64
Pedestrians Killed                  int64
Pedestrians Injured                 int64
Severity                           object
Intersection                       object
Alcohol Involved                   object
HazMat Involved                    object
Crash Type Code                    object
Total Vehicles Involved             int64
Crash Location                     object
Location Direction                 object
Route                              object
Route Suffix                       object
SRI (Std Rte Identifier)          

This shows us the type of data type object (or dtypes) the values of every column are. Objects refer to strings. Int64 are integers. Float64 are floats.

The warning at the beginning said it was column 6 & 17 that had mixed dtypes. If you look at your column list and count to the sixth column (Remember to count from zero!), you'll see that it's the 'Police Dept Code' column. Let's look at every unique value in that column.

In [183]:
njaccidents['Police Dept Code'].unique()

array(['01', '99', '  ', '02', '03', '04', '1', '2', '3', '4'], dtype=object)

And there it is! As you can see, there are strings and integers mixed together in the same column.

In [184]:
njaccidents['Crash Type Code'].unique()

array(['02', '08', '01', '06', '11', '05', '03', '13', '07', '99', '15',
       '14', '10', '04', '12', '09', '16', '  ', '00', '6', '2', '8', '3',
       '1', '7', '4', '5', '9'], dtype=object)

Same for column 17 or the 'Crash Type Code' column. Let's fix that by changing every value in both columns to a string using the astype() function.

In [185]:
njaccidents['Police Dept Code']=njaccidents['Police Dept Code'].astype(str)

In [186]:
njaccidents['Crash Type Code']=njaccidents['Crash Type Code'].astype(str)

We're changing it to a string because we don't need to do math with these numbers since they are codes so it's more beneficial to use them as objects. If you wanted to change something to an integer or a float, you'll need to use astype(int) and astype(float) respectively.

In [187]:
njaccidents['Police Dept Code'].unique()

array(['01', '99', '  ', '02', '03', '04', '1', '2', '3', '4'], dtype=object)

That took care of that!

Let's make our dataframe a little bit more manageable by weeding out some unnecessary columns. Let's also create a new dataframe called njcrashinfo.

In [188]:
njcrashinfo = njaccidents[['County Name', 'Municipality Name', 'Crash Date', 'Crash Day Of Week', 'Crash Time', 'Total Killed', 'Total Injured', 'Pedestrians Killed', 'Pedestrians Injured', 'Total Vehicles Involved', 'Alcohol Involved', 'Cell Phone In Use Flag']]

In [189]:
njcrashinfo

County Name         Municipality Name  Crash Date Crash Day Of Week  \
0          ATLANTIC  ABSECON CITY              03/04/2008                TU   
1          ATLANTIC  ABSECON CITY              12/19/2008                F    
2          ATLANTIC  ABSECON CITY              11/25/2008                TU   
3          ATLANTIC  ABSECON CITY              03/31/2008                M    
4          ATLANTIC  ABSECON CITY              01/25/2008                F    
5          ATLANTIC  ABSECON CITY              04/24/2008                TH   
6          ATLANTIC  ABSECON CITY              04/26/2008                SA   
7          ATLANTIC  ABSECON CITY              04/26/2008                SA   
8          ATLANTIC  ABSECON CITY              04/27/2008                S    
9          ATLANTIC  ABSECON CITY              11/24/2008                M    
10         ATLANTIC  ABSECON CITY              05/14/2008                W    
11         ATLANTIC  ABSECON CITY              01/01/2008                TU   
12         ATLANTIC  ABSECON CITY              01/02/2008                W    
13         ATLANTIC  ABSECON CITY              01/03/2008                TH   
14         ATLANTIC  ABSECON CITY              01/03/2008                TH   
15         ATLANTIC  ABSECON CITY              01/04/2008                F    
16         ATLANTIC  ABSECON CITY              01/05/2008                SA   
17         ATLANTIC  ABSECON CITY              01/05/2008                SA   
18         ATLANTIC  ABSECON CITY              01/08/2008                TU   
19         ATLANTIC  ABSECON CITY              01/08/2008                TU   
20         ATLANTIC  ABSECON CITY              01/08/2008                TU   
21         ATLANTIC  ABSECON CITY              01/08/2008                TU   
22         ATLANTIC  ABSECON CITY              01/10/2008                TH   
23         ATLANTIC  ABSECON CITY              01/10/2008                TH   
24         ATLANTIC  ABSECON CITY              01/10/2008                TH   
25         ATLANTIC  ABSECON CITY              01/11/2008                F    
26         ATLANTIC  ABSECON CITY              01/13/2008                S    
27         ATLANTIC  ABSECON CITY              01/14/2008                M    
28         ATLANTIC  ABSECON CITY              01/15/2008                TU   
29         ATLANTIC  ABSECON CITY              01/16/2008                W    
...             ...                       ...         ...               ...   
1707811      WARREN  WHITE TWP                 07/14/2013                S    
1707812      WARREN  WHITE TWP                 07/18/2013                TH   
1707813      WARREN  WHITE TWP                 07/20/2013                SA   
1707814      WARREN  WHITE TWP                 07/23/2013                TU   
1707815      WARREN  WHITE TWP                 07/26/2013                F    
1707816      WARREN  WHITE TWP                 07/27/2013                SA   
1707817      WARREN  WHITE TWP                 07/27/2013                SA   
1707818      WARREN  WHITE TWP                 08/01/2013                TH   
1707819      WARREN  WHITE TWP                 07/27/2013                SA   
1707820      WARREN  WHITE TWP                 08/06/2013                TU   
1707821      WARREN  WHITE TWP                 08/06/2013                TU   
1707822      WARREN  WHITE TWP                 08/06/2013                TU   
1707823      WARREN  WHITE TWP                 08/10/2013                SA   
1707824      WARREN  WHITE TWP                 08/15/2013                TH   
1707825      WARREN  WHITE TWP                 08/27/2013                TU   
1707826      WARREN  WHITE TWP                 08/28/2013                W    
1707827      WARREN  WHITE TWP                 08/30/2013                F    
1707828      WARREN  WHITE TWP                 08/31/2013                SA   
1707829      WARREN  WHITE TWP                 09/01/2013 

# How many car accidents had alcohol involved?

Let's find out the unique values that come up in the column 'Alcohol Involved'.

In [190]:
njcrashinfo['Alcohol Involved'].unique()

array(['N', 'Y'], dtype=object)

We have only two unique values in the column. N for no and Y for yes.

Let's find out how many incidents had Ns and how many had Ys. We're going to use the function value_counts() on the column 'Alcohol Involved'. We're also going to put the list in a new dataframe called alcohol so that it will look nicer in our notebook.

In [191]:
alcohol = pd.DataFrame(njcrashinfo['Alcohol Involved'].value_counts())

In [192]:
alcohol

Alcohol Involved
N           1658248
Y             49593

A lot more Ns than Ys. But just what percentage are the Ys compared to the Ns? First, let's get the total number of crashes in our data frame.

In [193]:
njcrashcount =njcrashinfo['Alcohol Involved'].count()

Careful. The count() function doesn't count NAs or null values. Always make sure to check for those using the isnull() function, followed by sum()

In [194]:
njcrashinfo['Alcohol Involved'].isnull().sum()

0

Let's create a new column named 'Percent' and divide every value of the 'Alcohol Involved' column by the total crashes from the 'njcrashcount' we created above and then multiply by 100.

In [195]:
alcohol['Percent'] = (alcohol['Alcohol Involved']/njcrashcount)*100

In [196]:
alcohol

Alcohol Involved    Percent
N           1658248  97.096158
Y             49593   2.903842

Mystery solved. Only 2.9 percent.

# How many total people were killed in every county?

Let's first use the value_counts() function to find out how many accidents were reported in each county.

In [197]:
njcrashinfo['County Name'].value_counts()

MIDDLESEX     176402
BERGEN        173913
ESSEX         163683
MONMOUTH      125018
UNION         117452
HUDSON        117339
PASSAIC       113217
OCEAN          94576
CAMDEN         93880
MORRIS         81992
MERCER         81310
SOMERSET       72856
BURLINGTON     70170
ATLANTIC       54068
GLOUCESTER     45056
CUMBERLAND     28001
HUNTERDON      24056
SUSSEX         23951
WARREN         20961
CAPE MAY       19839
SALEM          10101
Name: County Name, dtype: int64

So let's split up every incident that happened in every county by using the groupby()

In [198]:
njcrashinfo.groupby('County Name')

That looks like it did nothing, but it actually DID split up the counties into their own seperate groups. We just need to know perform an action. If you notice, there are columns like 'Total Killed', 'Total Injured', 'Pedestrians Killed', etc. that have numbers or integers that can be summed up. Basically, we're going to add them all up by using the sum() function.

In [199]:
njcrashinfo.groupby('County Name').sum()

Total Killed  Total Injured  Pedestrians Killed  \
County Name                                                    
ATLANTIC              187          20957                  33   
BERGEN                166          47575                  70   
BURLINGTON            249          23727                  44   
CAMDEN                230          36056                  62   
CAPE MAY               48           6114                   7   
CUMBERLAND            143          11086                  21   
ESSEX                 247          58229                  95   
GLOUCESTER            156          15964                  30   
HUDSON                122          29470                  44   
HUNTERDON              55           5835                  10   
MERCER                141          22701                  39   
MIDDLESEX             308          52131                  88   
MONMOUTH              230          36027                  52   
MORRIS                128          21039                  26   
OCEAN                 284          30520                  65   
PASSAIC               160          35694                  50   
SALEM                  72           3582                   4   
SOMERSET              119          19076                  17   
SUSSEX                 65           7004                   6   
UNION                 182          37662                  66   
WARREN                 69           6255                   6   

             Pedestrians Injured  Total Vehicles Involved  
County Name                                                
ATLANTIC                    1104                    98713  
BERGEN                      2879                   333025  
BURLINGTON                   634                   128223  
CAMDEN                      1539                   178112  
CAPE MAY                     258                    35792  
CUMBERLAND                   346                    49772  
ESSEX                       5224                   315666  
GLOUCESTER                   475                    81242  
HUDSON                      4026                   230722  
HUNTERDON                     95                    40153  
MERCER                       995                   153293  
MIDDLESEX                   2055                   336919  
MONMOUTH                    1239                   231862  
MORRIS                       658                   150796  
OCEAN                       1048                   176583  
PASSAIC                     2054                   216166  
SALEM                         55                    16512  
SOMERSET                     630                   133774  
SUSSEX                       144                    38997  
UNION                       2142                   227130  
WARREN                       154                    35105

Well, that's grim. Let's just take out the 'Total Killed' column using iloc which asks what data we should slice by putting an integer based on its position. The first value represents the rows and is separated by comma from the second value which represents the columns. Therefore, if we want all of the rows, we put a colon. We then seperate using a comma. Then, because 'Total Killed' is the first column, we can slice it by putting in a zero. We will also sort it by using sort_values and adding the option 'ascending=False' because we want the values to descend. Let's make it into a new dataframe called countydeaths.

In [200]:
countydeaths = njcrashinfo.groupby('County Name').sum().iloc[:,0].sort_values(ascending=False)

Let's make 'countydeaths' into a dataframe.

In [201]:
pd.DataFrame(countydeaths)

Total Killed
County Name              
MIDDLESEX             308
OCEAN                 284
BURLINGTON            249
ESSEX                 247
CAMDEN                230
MONMOUTH              230
ATLANTIC              187
UNION                 182
BERGEN                166
PASSAIC               160
GLOUCESTER            156
CUMBERLAND            143
MERCER                141
MORRIS                128
HUDSON                122
SOMERSET              119
SALEM                  72
WARREN                 69
SUSSEX                 65
HUNTERDON              55
CAPE MAY               48

# What about dates?

You may have noticed that the dates on the 'Crash Date' are strings and not Python date objects. This is bad because if you sort them you'll get '01/01/2008, 01/01/2009, 01/01/2010' etc. We want them to sort by date correctly, and in order to do that, we need to turn them into the Python date format. We will need to import datetime first. Then we will use apply() along with the lambda function to turn every string in that column into the format "%m/%d/%Y".

In [202]:
njcrashinfo['Crash Date']=njcrashinfo['Crash Date'].apply(lambda x: datetime.strptime(x, "%m/%d/%Y").date())

/Users/castudi/anaconda/envs/nicarpandas/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Now we're ready to groupby() the 'Crash Date' column every date in our dataframe and count how many accidents happened every day. And then we will slice the first column which is how many crashes happened each day using iloc. (Colon for all rows, comma, then 0 for the first column)

In [210]:
crashesbydate = njcrashinfo.groupby('Crash Date').count().iloc[:,0]

Now let's sort.

In [212]:
crashesbydate.sort_values(ascending=False)

Crash Date
2008-02-12    3050
2010-01-28    2180
2010-12-26    2017
2008-12-19    1995
2008-02-22    1893
2009-12-31    1791
2011-01-26    1767
2008-12-24    1655
2010-12-16    1644
2011-10-29    1535
2012-01-21    1522
2008-05-09    1480
2009-02-03    1459
2008-09-26    1400
2008-05-16    1375
2008-10-28    1366
2013-02-08    1361
2011-11-22    1356
2009-06-05    1351
2008-02-01    1351
2008-12-11    1335
2011-01-07    1334
2008-11-14    1326
2008-04-28    1312
2012-11-07    1308
2012-10-19    1308
2009-01-28    1299
2009-01-19    1296
2011-09-23    1293
2010-11-04    1293
              ... 
2013-02-17     450
2012-01-08     447
2009-03-08     446
2011-09-11     443
2013-02-24     441
2013-01-13     440
2008-12-25     440
2013-12-01     440
2009-11-26     439
2011-02-20     438
2012-01-29     436
2011-02-27     435
2013-03-24     432
2013-11-28     432
2013-11-10     431
2013-01-20     423
2013-03-03     418
2011-03-13     417
2009-01-04     416
2013-03-31     413
2012-12-25     405
2

Looks like on February 12, 2008 was a busy day for NJDOT with 3,050 accidents reported to happen that day.